### PR TITLE
Adds const iterators to AutoPas

### LINKED ISSUE #250: AutoPas and const correctness
**Describe the bug**
Currently, AutoPas doesn't handle const correctness at all.
E.g. using a `const AutoPas&` will result in more or less all interfaces disappearing because of the missing const keyword after the function declaration.

**Subtasks**
- [x] Put const after all getters in:
  - [x] the AutoPas class.
  - [x] the Container classes.
  - [x] Particle classes.
- [x] Implement const_iterator(s). This is probably best done using a template for the ParticleIterator's, which then point to const particles and don't allow deletion

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

This PR adds const iterators to AutoPas.
In contrast to normal iterators, they
- return a reference/pointer to `const Particle` instead of `Particle`
- and do not allow deletion of particles.

For the disabling of the deletion, the `deleteCurrentParticle` method had to be split in a template function that lives in the uppermost interface (ParticleIteratorInterface) and the respective implementation (`deleteCurrentParticleImpl`), which could not be disabled, as it is a virtual function.

## Additional Things
- Removes MoleculeLJ.cpp, as it is no longer needed due to inline static variables (c++17)

## Related Pull Requests

- Merge after #349, as this PR includes changes from it. (I didn't want any merge conflicts)

## Resolved Issues

- [x] fixes #250 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] Compiles
- [x] Additional test of particle iteration
- [x] Tested that using the deletion does not compile

## Subtasks from #250 
- Put const after all getters in:
  - [x] the AutoPas class.
  - the Container classes:
    - [x] PCInterface
    - [x] PC
    - [x] DS
    - [x] LC
    - [x] VLLB
    - [x] VL
    - [x] VarVL
    - [x] VLC
    - [x] VCL
  - Particle classes:
    - [x] ParticleBase
    - [x] SPHParticle
    - [x] MoleculeLJ
- [x] Implement const_iterator(s). This is probably best done using a template for the ParticleIterator's, which then point to const particles and don't allow deletion

## Notes
- For now the function `VerletLists::checkNeighborListsAreValid()` cannot be made const, since for VL a functor is called and we do not (yet) support const_functors (and probably never will)

### CODE REVIEW

**Comment by SteffenSeckler** [comment]
slightly better error messages are possible:

for
```C++
template <typename Dummy = void>
inline std::enable_if_t<modifiable, Dummy> deleteCurrentParticle() {
  deleteCurrentParticleImpl();
}
```

there exists an alternative implementation with potentially more meaningful error messages:

```C++
// in some utils header file:
template <typename T>
struct dependent_false {
  static constexpr bool value = false;
};

template <typename T = void>
inline void deleteCurrentParticle() {
  if constexpr (modifiable) {
    deleteCurrentParticleImpl();
  } else {
    static_assert(dependent_false<T>::value,
                  "You are trying to use deleteCurrentParticle on a const iterator, which is forbidden!");
  }
}
```
the error message will then look like:
```
/work/seckler/Documents/AutoPas/src/autopas/.././autopas/iterators/ParticleIteratorInterface.h:72:7: error: static_assert failed due to requirement 'dependent_false<void>::value' "You are trying to use deleteCurrentParticle on a const iterator, which is forbidden!"
      static_assert(dependent_false<T>::value,
      ^             ~~~~~~~~~~~~~~~~~~~~~~~~~
/work/seckler/Documents/AutoPas/tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:252:10: note: in instantiation of function template specialization 'autopas::ParticleIteratorInterface<autopas::MoleculeLJ, false>::deleteCurrentParticle<void>' requested here
    iter.deleteCurrentParticle();
         ^
```
instead of:
```
/work/seckler/Documents/AutoPas/tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:252:10: error: no matching member function for call to 'deleteCurrentParticle'
    iter.deleteCurrentParticle();
    ~~~~~^~~~~~~~~~~~~~~~~~~~~
/work/seckler/Documents/AutoPas/src/autopas/.././autopas/iterators/ParticleIteratorInterface.h:63:46: note: candidate template ignored: requirement 'false' was not satisfied [with Dummy = void]
  inline std::enable_if_t<modifiable, Dummy> deleteCurrentParticle() {
```

File: src/autopas/AutoPas.h [hunk 1]
```diff
@@ -178,12 +178,28 @@ class AutoPas {
...
-  autopas::ParticleIteratorWrapper<Particle> begin(IteratorBehavior behavior = IteratorBehavior::haloAndOwned) {
+  autopas::ParticleIteratorWrapper<Particle, true> begin(IteratorBehavior behavior = IteratorBehavior::haloAndOwned) {
     return _logicHandler->begin(behavior);
   }
 
+  /**
+   * @copydoc begin()
+   * @note const version
+   */
+  autopas::ParticleIteratorWrapper<Particle, false> begin(
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/AutoPas.h]: well, renaming does not make sense. As of STL, one could provide `cbegin` in addition to the two `begin` functions, see, e.g. std::vector, which provides the following three functions:
```C++
iterator begin() noexcept;
const_iterator begin() const noexcept;
const_iterator cbegin() const noexcept;
```

the difference between `begin` and `cbegin` being that cbegin will guarantee to return a `const_iterator`, while `begin` depends on the constness of the object.

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/AutoPas.h]: I have now _added_ cbegin, even though a const iterator can also be easily obtained using:
```C++
auto iter = std::as_const(obj).begin();
```

File: src/autopas/AutoPas.h [hunk 2]
```diff
@@ -178,12 +178,30 @@ class AutoPas {
    * for(auto iter = autoPas.begin(); iter.isValid(); ++iter)
    * @param behavior the behavior of the iterator. You can specify whether to iterate over owned particles, halo
    * particles, or both.
-   * @return iterator to the first particle
+   * @return iterator to the first particle_logicHandler
```
Comment (FG-TUM) [inline src/autopas/AutoPas.h]:
```suggestion
   * @return iterator to the first particle.
```

**Pull request merged**